lms/djangoapps/instructor/tests/test_certificates.py

```diff
@@ -734,3 +736,159 @@ def test_generate_certificate_exceptions_invalid_user_list_error(self):
+                      "test_student@example.com,test,1"
+
+        uploaded_file = SimpleUploadedFile("temp.csv", csv_content)
+        response = self.client.post(self.url, {'students_list': uploaded_file})
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.content)
+        self.assertEquals(len(data['errors']), 2)
+        self.assertEquals(len(data['warnings']), 0)
+        self.assertEquals(len(data['success']), 0)
+        self.assertEqual(data['errors'][0].get('response'), u'-- Data in row# 1 must have exactly two columns, either '
+                                                            u'an email or a username and notes.')
+
+    def test_file_upload_type_not_csv(self):
+        """
+        Try uploading some non-CSV file and verify that it is rejected
+        """
+        uploaded_file = SimpleUploadedFile("temp.jpg", io.BytesIO(b"some initial binary data: \x00\x01").read())
+        response = self.client.post(self.url, {'students_list': uploaded_file})
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.content)
+        self.assertNotEquals(len(data['errors']), 0)
+        self.assertEquals(data['errors'][0]['response'], '-- Make sure that the file you upload is in CSV format with '
+                                                         'no extraneous characters or rows.')
+
+    def test_bad_file_upload_type(self):
+        """
+        Try uploading some non-CSV file and verify that it is rejected
+        """
+        uploaded_file = SimpleUploadedFile("temp.csv", io.BytesIO(b"some initial binary data: \x00\x01").read())
+        response = self.client.post(self.url, {'students_list': uploaded_file})
+        self.assertEqual(response.status_code, 200)
+        data = json.loads(response.content)
+        self.assertNotEquals(len(data['errors']), 0)
+        self.assertEquals(data['errors'][0]['response'], '-- Could not read uploaded file.')
+
+    def test_invalid_email_in_csv(self):
+        """
+        Test failure case of a poorly formatted email field
+        """
+        csv_content = "test_student.example.com,dummy_notes"
+
+        uploaded_file = SimpleUploadedFile("temp.csv", csv_content)
+        response = self.client.post(self.url, {'students_list': uploaded_file})
+        data = json.loads(response.content)
+        self.assertEqual(response.status_code, 200)
+        self.assertNotEquals(len(data['errors']), 0)
+        self.assertEquals(len(data['warnings']), 0)
+        self.assertEquals(data['errors'][0]['response'], u'-- user (username/email={0}) does not exist.'.
+                          format('test_student.example.com'))
+
+    def test_csv_user_not_enrolled(self):
+        """
+        If the user is not enrolled in the course then there should be a warning.
+        """
+        csv_content = "nonenrolled@test.com,dummy_notes"
+
+        uploaded_file = SimpleUploadedFile("temp.csv", csv_content)
```

Comment:
Can we make a private method for these tests having these statements to avoid repetition of code?
```
uploaded_file = SimpleUploadedFile("temp.csv", csv_content)
response = self.client.post(self.url, {'students_list': uploaded_file})
self.assertEqual(response.status_code, 200)
data = json.loads(response.content)
```